### DISCUSSION TITLE
Installation on NixOS

### DISCUSSION THREAD

**Discussion by viperML** [glpi-project/glpi-agent#689] [General]
### Your question

Hello, 

I need to install this software on NixOS, for which I am writing the package and system configuration here: https://github.com/NixOS/nixpkgs/compare/master...viperML:nixpkgs:glpi

Right now glpi-agent is trying to write its var data into /nix/store/----glpi-agent/var , whic is not right in this distribution. I have tried configuring /etc/glpi-agent/conf.d/nixos.cfg to include `vardir = /var/lib/glpi-agent`, but doesn't seem to be applied. I don't know how to debug this further.

It is possible to create and run a NixOS VM, having a a nix installation and running the following command on the root of my nixpkgs fork:

```
$ nix build -I nixos-config=$PWD/configuration.nix -I nixpkgs=$PWD -f ./nixos config.system.build.vm -L
$ ./result/bin/run-nixos-vm
```

**Comment by g-bougard** [comment]
Hello @viperML 

glad to here about a new packaging support.

I really don't know anything about NixOS, but I can give you few advice as I worked on debian, rpm, MacOSX & Windows packaging ;-)

About `vardir`, I still can advise you to not set it in config for packages, but in the `setup.pm` file. This is where it should be set the default `vardir` in packages. Sometime you may want to generate the `setup.pm` file yourself, but you can use the "prepare" step, see below.

So you can also "prepare" the source running `perl Makefile.pl` at sometime. This preparation step accepts some variables. You can see exemples in the [rpm spec](https://github.com/glpi-project/glpi-agent/blob/d8b857e21d017f5138e8780c53b79ad8a70e0226/contrib/unix/glpi-agent.spec#L129) file, in the "%build" section. Also there in the "%prep" section, you can see the default "agent.cfg" file is modified to better match what could be expected by default on a given OS. For the debian package, you have the sample of variables for `perl Makefile.PL` here: [debian/rules](https://github.com/glpi-project/glpi-agent/blob/d8b857e21d017f5138e8780c53b79ad8a70e0226/debian/rules#L16). You will see you can set `vardir` setting `LOCALSTATEDIR` with `perl Makefile.PL`.

I hope this helps.

**Reply by viperML** [reply]
Hello, thank you for your insight. After having configured LOCALSTATEDIR, it seems to work on a first approach.

One of the big differences in NixOS is that packages are installed into a unique prefix `/nix/store/....name` instead of `/usr`. So I have chose to configure `SYSCONFDIR=/nix/store/....glpi-agent/etc/glpi-agent`, and changed the `agent.conf` to read the user-supplied configuration from `/etc/glpi-agent/conf.d`. Does this sound sane?

**Reply by g-bougard** [reply]
Hi @viperML 

if this works, that's just fine. Good job.

It's always better to follow installed OS philosophy.

**Comment by liberodark** [comment]
This would be fix here : https://github.com/NixOS/nixpkgs/issues/361741

**Reply by g-bougard** [reply]
Nice, thank you.
